### PR TITLE
[prebuilds] Introduce 'failed' state for prebuilds

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

This PR introduces a "failure" state for prebuilds that have entirely failed to finish.
This is to distinguish them from failed tasks but finished rebuilds.

## Related Issue(s)
fixes https://github.com/gitpod-io/gitpod/issues/8592

## Release Notes

```release-note
NONE
```

### CODE REVIEW

File: components/dashboard/src/projects/Prebuilds.tsx [hunk 1]
```diff
@@ -248,6 +248,8 @@ export function prebuildStatusLabel(prebuild?: PrebuildWithStatus) {
             return (<span className="font-medium text-blue-500 uppercase">running</span>);
         case "aborted":
             return (<span className="font-medium text-gray-500 uppercase">canceled</span>);
+        case "failed":
```
Comment (jankeromnes) [inline components/dashboard/src/projects/Prebuilds.tsx:251]: Drive-by comment: You probably want to handle `"failed"` in `prebuildStatusIcon` below as well (there is already a `StatusFailed` icon you can re-use).

**Review by JanKoehnlein — APPROVED** [review]
LGTM

File: components/ws-manager-bridge/ee/src/bridge.ts [hunk 2]
```diff
@@ -89,6 +89,9 @@ export class WorkspaceManagerBridgeEE extends WorkspaceManagerBridge {
                     prebuild.error = status.conditions!.headlessTaskFailed;
                     prebuild.snapshot = status.conditions!.snapshot;
                     headlessUpdateType = HeadlessWorkspaceEventType.FinishedButFailed;
+                } else if (!status.conditions!.snapshot) {
+                    prebuild.state = "failed";
```
Comment (jankeromnes) [inline components/ws-manager-bridge/ee/src/bridge.ts:93]: Suggestion: Here you could also set `prebuild.error` to something like "Prebuild does not have a snapshot".

File: components/dashboard/src/projects/Prebuilds.tsx (hunk 1 above)
Comment (easyCZ) [reply components/dashboard/src/projects/Prebuilds.tsx:251]: As the number of error cases grows, it becomes harder to enumerate them. One possible way to simplify the API UX is to have a single `error` state and extend it with a `cause` which is set to a computer readable reason for the failure - TimedOut, RateLimited, FailedToStart, etc. The error message then becomes a human readable string to show in the UI etc.

Comment (jankeromnes) [reply components/dashboard/src/projects/Prebuilds.tsx:251]: Good point @easyCZ. Alternatively, I think we could also special-case the few non-error cases, then consider "all the rest" as errors?

**Comment by svenefftinge** [comment]
wow, that was quick feedback 😁 The branchname was too long so I had to crate [another PR](https://github.com/gitpod-io/gitpod/pull/8602).

File: components/ws-manager-bridge/ee/src/bridge.ts (hunk 2 above)
Comment (svenefftinge) [reply components/ws-manager-bridge/ee/src/bridge.ts:93]: That is a symptom not the actual error. I've discussed with Gero that we will also relay overwriting actual errors if they exist (in case of multiple stop messages).

File: components/dashboard/src/projects/Prebuilds.tsx (hunk 1 above)
Comment (easyCZ) [reply components/dashboard/src/projects/Prebuilds.tsx:251]: Depends how far we want to go. Some failures are due to user issues, some can be due to system issues. I think the it mostly depends if we control all the `clients` which will be interpreting the status (Dashboard, etc) or if some of the clients are/will be out of our control (implementing against an API). In the former case, it doesn't matter much as we can largely reach a shared understanding of what is an error and what isn't. But if an external frontend/server needs to `switch` on these, it may be less obivous to them with a large number of cases.

I think this is probs not something we need/should decide now and I don't think it blocks the change.
